### ISSUE TITLE
lotus-storage-miner blocks on start

### ISSUE THREAD

**Issue by dtynn** [filecoin-project/lotus#1126]
**Describe the bug**
lout-storage-miner blocked on start
you can reproduce it by just initializing a private net, and it probably (30%~50%) happens

**To Reproduce**
Steps to reproduce the behavior on private net

1. Initializing 

**for lotus**
```
# prepare pre-seal sectors
lotus-seed --sectorbuilder-dir=./repo/genesis-sectors pre-seal --num-sectors=3

# start lotus daemon
lotus daemon --lotus-make-random-genesis=./genesis.car --bootstrap=false --genesis-presealed-sectors=./repo/genesis-sectors/pre-seal-t0101.json
```

**for miner**
```
# init miner
lotus-storage-miner init --actor=t0101 --genesis-miner --nosync --pre-sealed-sectors=./repo/genesis-sectors/

# start miner
lotus-storage-miner run --nosync
```

2. See error
miner instance blocks

if miner is interrupted (Ctrl-C), lotus output logs:
```
2020-01-21T08:31:13.246+0800    INFO    badger  v2@v2.0.1/logger.go:46  Replay took: 221.299µs

2020-01-21T08:31:13.246+0800    INFO    p2pnode lp2p/addrs.go:114       Swarm listening at: [/ip4/127.0.0.1/tcp/63939 /ip4/192.168.0.69/tcp/63939 /ip6/::1/tcp/63940]
2020-01-21T08:31:13.247+0800    WARN    hello   hello/hello.go:50       running without peer manager
panic: send on closed channel

goroutine 375 [running]:
github.com/filecoin-project/lotus/lib/jsonrpc.(*wsConn).handleChanOut(0xc000358ab0, 0x551eea0, 0xc01d9acee8, 0x92, 0x16, 0x8)
        /Users/wanglin/work/ongit/lotus/lib/jsonrpc/websocket.go:209 +0xa6
github.com/filecoin-project/lotus/lib/jsonrpc.handlers.handle.func1(0xb951530, 0xc01da68cc0)
        /Users/wanglin/work/ongit/lotus/lib/jsonrpc/handler.go:232 +0xde
github.com/filecoin-project/lotus/lib/jsonrpc.(*wsConn).nextWriter(0xc000358ab0, 0xc0004da640)
        /Users/wanglin/work/ongit/lotus/lib/jsonrpc/websocket.go:111 +0x1ba
github.com/filecoin-project/lotus/lib/jsonrpc.handlers.handle(0xc00020c1e0, 0x5a86fa0, 0xc01d9cd980, 0xc01da5c110, 0x3, 0xc01da5c118, 0xc000045d40, 0x14, 0x6576b08, 0x0, ...)
        /Users/wanglin/work/ongit/lotus/lib/jsonrpc/handler.go:226 +0xb3d
created by github.com/filecoin-project/lotus/lib/jsonrpc.(*wsConn).handleCall
        /Users/wanglin/work/ongit/lotus/lib/jsonrpc/websocket.go:362 +0x295
```

**Expected behavior**
miner begin mining after a while

**Version (run `lotus --version`):**
lotus version 0.2.5

**Additional context**
seems to be some mutex problem, tried to profile it, and add the codes in `cmd/lotus/rpc.go`
```
func init() {
	runtime.SetBlockProfileRate(1000)
	runtime.SetMutexProfileFraction(5)
}

```

the bug just disappears(or more difficult to reproduce?)

**Comment by dtynn** [comment]
1. master branch
![image](https://user-images.githubusercontent.com/1426666/72766933-05663280-3c2d-11ea-817e-936c79ac4d73.png)

2. miner blocks
![image](https://user-images.githubusercontent.com/1426666/72766961-1fa01080-3c2d-11ea-8b7f-247a225b323c.png)

3. lotus crashes
![image](https://user-images.githubusercontent.com/1426666/72766973-347ca400-3c2d-11ea-8df0-77d21ca94c3c.png)

4. add profile codes
![image](https://user-images.githubusercontent.com/1426666/72767022-786fa900-3c2d-11ea-833b-83c14dba27fd.png)
![image](https://user-images.githubusercontent.com/1426666/72767050-9ccb8580-3c2d-11ea-94de-8df6a08cec51.png)

5. use the changed binary
![image](https://user-images.githubusercontent.com/1426666/72767097-dd2b0380-3c2d-11ea-8a65-24ca7a634d5b.png)

6. rollback to original 
![image](https://user-images.githubusercontent.com/1426666/72767195-35fa9c00-3c2e-11ea-912f-4727cbf16a46.png)


7. miner block again
![image](https://user-images.githubusercontent.com/1426666/72767282-95f14280-3c2e-11ea-9782-a36caf6fec85.png)

**Comment by dtynn** [comment]
`go version go1.13.3 darwin/amd64`

**Comment by whyrusleeping** [comment]
The lotus panic appears to be a bug in the jsonrpc library... will have to have @magik6k look into that.

As for the hang, i'm really not sure... I will try to reproduce locally.

**Comment by dtynn** [comment]
@whyrusleeping  
maybe #1122 is duplicate with this issue

and create a work around pr #1129 

this bug is actually caused by `chan handling along with some mutex usage` in the `lib/jsonrpc`, but hard to finally locate and fix

**Comment by hunjixin** [comment]
deadlock！hang up because of condition race between writeLk and registerCh.

assume one request is doing nextwrite， writelk is in lock，and attempt to write to registerCh，another requst may wait for writeLk in sendrequest function and block handleOutChans loop，  no one will select registerCh in this condition . eventually, a deadlock occurred.

```go
func (c *wsConn) sendRequest(req request) {            
	c.writeLk.Lock()                                   
	if err := c.conn.WriteJSON(req); err != nil {      
		log.Error("handle me:", err)                   
		c.writeLk.Unlock()                             
		return                                         
	}                                                  
	c.writeLk.Unlock()                                 
}                                                      
```

```go
func (c *wsConn) nextWriter(cb func(io.Writer)) {        
	c.writeLk.Lock()                                     
	defer c.writeLk.Unlock()                             
                                                         
	wcl, err := c.conn.NextWriter(websocket.TextMessage) 
	if err != nil {                                      
		log.Error("handle me:", err)                     
		return                                           
	}                                                    
                                                         
	cb(wcl)                                              
                                                         
	if err := wcl.Close(); err != nil {                  
		log.Error("handle me:", err)                     
		return                                           
	}                                                    
}                                                        
```

```go
func (c *wsConn) handleOutChans() {
	regV := reflect.ValueOf(c.registerCh)

	cases := []reflect.SelectCase{
		{ // registration chan always 0
			Dir:  reflect.SelectRecv,
			Chan: regV,
		},
	}
	var caseToID []uint64

	for {
		chosen, val, ok := reflect.Select(cases)
                .......
		if !ok {
			// Output channel closed, cleanup, and tell remote that this happened

			n := len(caseToID)
			if n > 0 {
				cases[chosen] = cases[n]
				caseToID[chosen-1] = caseToID[n-1]
			}

			id := caseToID[chosen-1]
			cases = cases[:n]
			caseToID = caseToID[:n-1]

			c.sendRequest(request{
				Jsonrpc: "2.0",
				ID:      nil, // notification
				Method:  chClose,
				Params:  []param{{v: reflect.ValueOf(id)}},
			})
			continue
		}

		// forward message
		c.sendRequest(request{
			Jsonrpc: "2.0",
			ID:      nil, // notification
			Method:  chValue,
			Params:  []param{
				{v: reflect.ValueOf(caseToID[chosen-1])},
				{v: val}  ,
			},
		})
		fmt.Println("send request finish")
	}
}
```

**Comment by dtynn** [comment]
ok, maybe here is the reason:

assuming 2 concurrent ChainNotify call:

-   first call expand the selection, get a HCCurrent immediately, and try to send it out, which will acquire the write lock
https://github.com/filecoin-project/lotus/blob/master/lib/jsonrpc/websocket.go#L193-L198

-  the second call finds a chan-kind return value, hold the write lock and try to send a registration via the registerCh
https://github.com/filecoin-project/lotus/blob/master/lib/jsonrpc/handler.go#L232

- since there is no receiver of the registerCh, and no buffer, boom

so, as a quick fix, you can just add some buffer size here:
https://github.com/filecoin-project/lotus/blob/master/lib/jsonrpc/websocket.go#L391

but i think it should be more carefully handled

**Comment by magik6k** [comment]
Thanks for investigating, I may have a fix for this

**Comment by magik6k** [comment]
Yeah, that deadlock was pretty fun to resolve - pushed the fix to https://github.com/filecoin-project/lotus/pull/1123/commits/fbc0330fa817a36d14e7063531538289e136d545, can you verify that it solves your issues?

**Comment by dtynn** [comment]
i think it is fixed in #1123

**Comment by magik6k** [comment]
#1123 was merged
